core/src/avm2/globals/avmplus.rs

```diff
@@ -1 +1,519 @@
+    activation: &mut Activation<'_, 'gc>,
+    _this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let value = args[0].coerce_to_object(activation)?;
+    let flags = DescribeTypeFlags::from_bits(args.get_u32(activation, 1)?).expect("Invalid flags!");
+
+    let class_obj = value.as_class_object().or_else(|| value.instance_of());
+    let mut object = activation
+        .avm2()
+        .classes()
+        .object
+        .construct(activation, &[])?;
+    let Some(class_obj) = class_obj else {
+        return Ok(Value::Null);
+    };
+
+    let is_static = value.as_class_object().is_some();
+    if !is_static && flags.contains(DescribeTypeFlags::USE_ITRAITS) {
+        return Ok(Value::Null);
+    }
+
+    let class = class_obj.inner_class_definition();
+    let class = class.read();
+
+    let qualified_name = class
+        .name()
+        .to_qualified_name(activation.context.gc_context);
+
+    object.set_public_property("name", qualified_name.into(), activation)?;
+
+    object.set_public_property(
+        "isDynamic",
+        (is_static || !class.is_sealed()).into(),
+        activation,
+    )?;
+    object.set_public_property(
+        "isFinal",
+        (is_static || class.is_final()).into(),
+        activation,
+    )?;
+    object.set_public_property("isStatic", is_static.into(), activation)?;
+
+    let traits = describe_internal_body(activation, class_obj, is_static, flags)?;
+    object.set_public_property("traits", traits.into(), activation)?;
+
+    Ok(object.into())
+}
+
+bitflags::bitflags! {
+    pub struct DescribeTypeFlags: u32 {
+        const HIDE_NSURI_METHODS      = 1 << 0;
+        const INCLUDE_BASES           = 1 << 1;
+        const INCLUDE_INTERFACES      = 1 << 2;
+        const INCLUDE_VARIABLES       = 1 << 3;
+        const INCLUDE_ACCESSORS       = 1 << 4;
+        const INCLUDE_METHODS         = 1 << 5;
+        const INCLUDE_METADATA        = 1 << 6;
+        const INCLUDE_CONSTRUCTOR     = 1 << 7;
+        const INCLUDE_TRAITS          = 1 << 8;
+        const USE_ITRAITS             = 1 << 9;
+        const HIDE_OBJECT             = 1 << 10;
+    }
+}
+
+fn describe_internal_body<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    class_obj: ClassObject<'gc>,
+    is_static: bool,
+    flags: DescribeTypeFlags,
+) -> Result<Object<'gc>, Error<'gc>> {
+    // If we were passed a non-ClassObject, or the caller specifically requested it, then
+    // look at the instance "traits" (our implementation is different than avmplus)
+
+    let use_instance_traits = !is_static || flags.contains(DescribeTypeFlags::USE_ITRAITS);
+    let mut traits = activation
+        .avm2()
+        .classes()
+        .object
+        .construct(activation, &[])?;
+
+    let bases = activation
+        .avm2()
+        .classes()
+        .array
+        .construct(activation, &[])?
+        .as_array_object()
+        .unwrap();
+
+    let interfaces = activation
```

Comment:
we generally write this via `ArrayObject::from_storage(activation, ArrayStorage::new(0)`, no need to go through the constructor/allocator dispatch :D

Reply:
We could maybe wanna wrap it in `ArrayObject::empty(acativation)`?

(Alternatively, (though I don't think you'll prefer that - just throwing as a possibility), you could start with `Storage`s, fill them and create the `ArrayObject`s at the end.)

Reply:
we could do a similar thing for `classes().object` (tbh I'm surprised we didn't have it before):

```rs
impl<'gc> ScriptObject<'gc> {
    pub fn new<'gc>(
        activation: &mut Activation<'_, 'gc>,
    ) -> Object<'gc> {
        let class = self.avm2().classes().object;
        ScriptObject(GcCell::new(activation.context.gc_context, ScriptObjectData::new(class))).into()
    }
```